context_test.go

```diff
@@ -0,0 +1,244 @@
+		NewResponse(400, "desc").Model(&ErrorModel{}),
+	).Run(func(ctx Context) {
+		ctx.AddError(fmt.Errorf("some error"))
+		ctx.AddError(&ErrorDetail{
+			Message:  "Invalid value",
+			Location: "body.field",
+			Value:    "test",
+		})
+		ctx.WriteError(http.StatusBadRequest, "test error")
+	})
+
+	var parsed ErrorModel
+	expected := ErrorModel{
+		Status: http.StatusBadRequest,
+		Title:  http.StatusText(http.StatusBadRequest),
+		Detail: "test error",
+		Errors: []*ErrorDetail{
+			{Message: "some error"},
+			{
+				Message:  "Invalid value",
+				Location: "body.field",
+				Value:    "test",
+			},
+		},
+	}
+
+	// No preference
+	w := httptest.NewRecorder()
+	req, _ := http.NewRequest(http.MethodGet, "/error", nil)
+	app.ServeHTTP(w, req)
+
+	assert.Equal(t, http.StatusBadRequest, w.Code)
+	assert.Equal(t, "application/problem+json", w.Header().Get("Content-Type"))
+	err := json.Unmarshal(w.Body.Bytes(), &parsed)
+	assert.NoError(t, err)
+	assert.Equal(t, expected, parsed)
+
+	// Prefer JSON
+	w = httptest.NewRecorder()
+	req, _ = http.NewRequest(http.MethodGet, "/error", nil)
+	req.Header.Set("Accept", "application/json")
+	app.ServeHTTP(w, req)
+
+	assert.Equal(t, http.StatusBadRequest, w.Code)
+	assert.Equal(t, "application/problem+json", w.Header().Get("Content-Type"))
+	err = json.Unmarshal(w.Body.Bytes(), &parsed)
+	assert.NoError(t, err)
+	assert.Equal(t, expected, parsed)
+
+	// Prefer YAML
+	w = httptest.NewRecorder()
+	req, _ = http.NewRequest(http.MethodGet, "/error", nil)
+	req.Header.Set("Accept", "application/yaml")
+	app.ServeHTTP(w, req)
+
+	assert.Equal(t, http.StatusBadRequest, w.Code)
+	assert.Equal(t, "application/problem+yaml", w.Header().Get("Content-Type"))
+	err = yaml.Unmarshal(w.Body.Bytes(), &parsed)
+	assert.NoError(t, err)
+	assert.EqualValues(t, expected, parsed)
+
+	// Prefer CBOR
+	w = httptest.NewRecorder()
+	req, _ = http.NewRequest(http.MethodGet, "/error", nil)
+	req.Header.Set("Accept", "application/cbor")
+	app.ServeHTTP(w, req)
+
+	assert.Equal(t, http.StatusBadRequest, w.Code)
+	assert.Equal(t, "application/problem+cbor", w.Header().Get("Content-Type"))
+	err = cbor.Unmarshal(w.Body.Bytes(), &parsed)
+	assert.NoError(t, err)
+	assert.Equal(t, expected, parsed)
+}
+
+func TestInvalidModel(t *testing.T) {
+	type R1 struct {
+		Foo string `json:"foo"`
+	}
+
+	type R2 struct {
+		Bar string `json:"bar"`
+	}
+
+	app := newTestRouter()
+
+	app.Resource("/bad-status").Get("test", "Test",
+		NewResponse(http.StatusOK, "desc").Model(R1{}),
+	).Run(func(ctx Context) {
+		ctx.WriteModel(http.StatusNoContent, R2{Bar: "blah"})
+	})
+
+	app.Resource("/bad-model").Get("test", "Test",
+		NewResponse(http.StatusOK, "desc").Model(R1{}),
+	).Run(func(ctx Context) {
+		ctx.WriteModel(http.StatusOK, R2{Bar: "blah"})
+	})
+
+	assert.Panics(t, func() {
+		w := httptest.NewRecorder()
+		req, _ := http.NewRequest(http.MethodGet, "/bad-status", nil)
+		app.ServeHTTP(w, req)
+	})
+
+	assert.Panics(t, func() {
```

Comment:
n/m this. I see we can do such a thing by applying the `Recovery` middleware.

Reply:
Good thinking. This actually isn't obvious in the test, but we are creating the router **without** the recovery middleware since you might write your own. That makes it easy to assert that it would panic in certain situations and makes writing the tests easier along with preventing circular imports. I'll try to add a comment about that.

In the middleware package there is a test to assert you do in fact get a 5xx, e.g. see:

```go
func TestRecoveryMiddleware(t *testing.T) {
	app, _ := newTestRouter(t)

	app.Resource("/panic").Get("panic", "Panic recovery test",
		responses.NoContent(),
	).Run(func(ctx huma.Context) {
		panic(fmt.Errorf("Some error"))
	})

	w := httptest.NewRecorder()
	req, _ := http.NewRequest(http.MethodGet, "/panic", nil)
	app.ServeHTTP(w, req)
	assert.Equal(t, http.StatusInternalServerError, w.Code)
	assert.Equal(t, "application/problem+json", w.Result().Header.Get("content-type"))
}
```

https://github.com/danielgtaylor/huma/pull/11/files#diff-ddd4a23b5d23539d33f99dcf0289088fR45